staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go

```diff
@@ -98,5 +110,17 @@ func (a *cachedTokenAuthenticator) AuthenticateToken(ctx context.Context, token
 }
 
 func keyFunc(auds []string, token string) string {
-	return fmt.Sprintf("%#v|%v", auds, token)
+	// individually hash each input to prevent ambiguities, i.e. "xy" + "z" == "x" + "yz"
+	audsHash := sha256.Sum256([]byte(fmt.Sprintf("%#v", auds)))
```

Comment:
consider defining:
```
func singleDirectHash(item interface{}) []byte {
  h := sha256.New()
  fmt.Fprintf(h, "%#v", item)
  return h.Sum()
}

Reply:
That should avoid one or more allocations.

Reply:
You can avoid the re-initializing the hash and Sprintf parsing with length prefixing.

```golang
func appendString(h hash.Hash, s string) {
  b := make([]byte, 4 + len(s))
  binary.LittleEndian.PutUint32(b[0:], uint32(len(s)))
  copy(b[4:], s)
  h.Write(b) 
}
```

This solves the ambiguous plaintext problem and also requires minimal allocated bytes. You could also allocate space for all strings in slice since it's easy to calculate total size.

There's also a nice wrapper for length prefix encoding that we use in the kms plugin: https://github.com/kubernetes/kubernetes/blob/a84787465547697106f43e30e4406bb490d02670/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/envelope.go#L135-L139.

Reply:
I went with the length prefix based approach since it resulted in the least number of allocations.